src/dialer/index.js

```diff
@@ -56,72 +58,112 @@ class Dialer {
+   * will be used.
    *
-   * @param {Array<Multiaddr>|Multiaddr} addrs
+   * @param {PeerInfo|Multiaddr} peer The peer to dial
    * @param {object} [options]
    * @param {AbortSignal} [options.signal] An AbortController signal
    * @returns {Promise<Connection>}
    */
-  async connectToMultiaddr (addrs, options = {}) {
-    if (!Array.isArray(addrs)) addrs = [multiaddr(addrs)]
-
-    const dialAction = (addr, options) => {
-      if (options.signal.aborted) throw errCode(new Error('already aborted'), 'ERR_ALREADY_ABORTED')
-      return this.transportManager.dial(addr, options)
+  async connectToPeer (peer, options = {}) {
+    const dialTarget = this._createDialTarget(peer)
+    if (dialTarget.addrs.length === 0) {
+      throw errCode(new Error('The dial request has no addresses'), 'ERR_NO_DIAL_MULTIADDRS')
     }
-    const dialRequest = new DialRequest({
-      addrs,
-      dialAction,
-      dialer: this
-    })
-
-    // Combine the timeout signal and options.signal, if provided
-    const timeoutController = new TimeoutController(this.timeout)
-    const signals = [timeoutController.signal]
-    options.signal && signals.push(options.signal)
-    const signal = anySignal(signals)
-
-    const dial = {
-      dialRequest,
-      controller: timeoutController
-    }
-    this._pendingDials.add(dial)
+    const pendingDial = this._pendingDials.get(dialTarget.id) || this._createPendingDial(dialTarget, options)
 
     try {
-      const dialResult = await dialRequest.run({ ...options, signal })
-      log('dial succeeded to %s', dialResult.remoteAddr)
-      return dialResult
+      const connection = await pendingDial.promise
+      log('dial succeeded to %s', dialTarget.id)
+      return connection
     } catch (err) {
       // Error is a timeout
-      if (timeoutController.signal.aborted) {
+      if (pendingDial.controller.signal.aborted) {
         err.code = codes.ERR_TIMEOUT
       }
       log.error(err)
       throw err
     } finally {
-      timeoutController.clear()
-      this._pendingDials.delete(dial)
+      pendingDial.destroy()
     }
   }
 
   /**
-   * Connects to a given `PeerInfo` or `PeerId` by dialing all of its known addresses.
-   * The dial to the first address that is successfully able to upgrade a connection
-   * will be used.
-   *
-   * @param {PeerId} peerId The remote peer id to dial
+   * @typedef DialTarget
+   * @property {string} id
+   * @property {Multiaddr[]} addrs
+   */
+
+  /**
+   * Creates a DialTarget. The DialTarget is used to create and track
+   * the DialRequest to a given peer.
+   * @private
+   * @param {PeerInfo|Multiaddr} peer A PeerId or Multiaddr
+   * @returns {DialTarget}
+   */
+  _createDialTarget (peer) {
+    const dialable = Dialer.getDialable(peer)
+    if (multiaddr.isMultiaddr(dialable)) {
+      return {
+        id: dialable.toString(),
+        addrs: [dialable]
+      }
+    }
+    // Check the peerstore first, and then fallback to the instance
```

Comment:
What do you mean by fallback here?

Reply:
I can remove that I think it's just residual from when I was moving stuff around and trying some things out. It will now just do a PeerStore update and get the full list of existing and new addrs.